keps/sig-storage/177-volume-snapshot-validation/tighten-validation-webhook-crd.md

```diff
@@ -0,0 +1,799 @@
+
+Admission controllers have been in common use in kubernetes for a long time. Admission webhooks are the new, preferred way to control admission, especially for external (out-of-tree) components like the CSI external snapshotter.
+
+A webhook server receives AdmissionReview(definition) requests from API server, and responses with a response of the same type to either admit/deny the request. Following simplified diagram shows the workflow.
+(Note that the mutatting webhooks will be invoked BEFORE validating).
+
+The webhook server will expose an HTTP endpoint such that to allow the API server to send AdmissionReview requests. Webhook server providers can dynamically(details) configure what type of resources and what type of admission webhooks via creating CRs of type ValidatingWebhookConfiguration and/or MutatingWebhookConfiguration. It’s worth thinking to provide a generic webhook server in csi-repo. At this stage, the proposal is to create a repo under CSI org, however only implement the validating webhooks for VolumeSnapshot.
+
+CRD validation is preferred over webhook validation due to their lower complexity, however CRD validation schema is unable to enforce immutability or provide ratcheting validation.
+
+### Goals
+
+- Provide an updated CRD schema to validate fields
+- Prevent:
+  - Invalid VolumeSnapshot/VolumeSnapshotContent from creation and update
+  - Invalid updates on immutable fields, i.e., VolumeSnapshot.Spec.Source
+- Provide a pre-built image which can be used to deploy the server
+- Provide a way to deploy the webhook server in cluster
+- Provide a way to authenticate the webhook server to the API server via TLS
+- Provide a release process to safely tighten the validation and move towards the ideal state of using builtin CRD validation while maintaining backwards compatibility
+
+### Non-Goals
+
+- Provide a way to authenticate the API server to the webhook server
+
+## Proposal
+
+Tighten the validation on Volume Snapshot objects. The following fields will begin to enforce immutability: `VolumeSnapshot.Spec.Source` and `VolumeSnapshotContent.Spec.Source`. The following fields will begin to enforce `oneOf`: `VolumeSnapshot.Spec.Source` and `VolumeSnapshotContent.Spec.Source`). The following fields will begin to enforce non-empty strings: `spec.VolumeSnapshotClassName`. More details are in the Validating Scenarios section.
+
+Due to backwards compatibility concerns, the tightening will occur in two phases.
+
+1. The first phase is webhook-only, and will use ratcheting validation combined with a reconciliation method to delete or fix currently persisted objects which are invalid under the new (strict) validation rules.
+2. The second phase occurs once all invalid objects are cleared from the cluster. The CRD schema validation will be tightened and the webhook will stick around to enforce immutability until immutable fields come to CRDs. (or the crd upgrade could wait until immutable fields are available to do in one go)
```

Comment:
Can you clarify here is the CRD schema going to be updated in the existing v1beta1? Or are we going to have a v1beta2 or do it in v1?

I personally prefer a v1beta2 or v1 because the API group version is the only version identifier we have. We can't really use K8s versions as an identifier since this CRD is installed outside of K8s.

Reply:
Since the validation is tightening is considered an api change which is non backwards compatible, we will probably need a v1beta2 or v1. Will we need a way to convert between the versions / validation? 

Will we need a new version for phase 1 or phase 2 or both?

Reply:
I think for phase 1(ratcheting validation), its OK to keep the original v1beta1 for CRD, and then phase 2 upgrades to V1? one thing to sort out is the feasibility of downgrade process in case a user needs to delete/update existing invalid CRs AFTER upgrading to V1.

Reply:
There will be a new version only for phase 2. It will not have any auto conversion except for changing the api version.